pkg/webhook/server.go

```diff
@@ -53,256 +50,111 @@ type ServerOptions struct {
-	manager manager.Manager
-
-	// httpServer is the actual server that serves the traffic.
-	httpServer *http.Server
-
-	once sync.Once
-}
-
-// Webhook defines the basics that a webhook should support.
-type Webhook interface {
-	// GetName returns the name of the webhook.
-	GetName() string
-	// GetPath returns the path that the webhook registered.
-	GetPath() string
-	// GetType returns the Type of the webhook.
-	// e.g. mutating or validating
-	GetType() types.WebhookType
-	// Handler returns a http.Handler for the webhook.
-	Handler() http.Handler
-	// Validate validates if the webhook itself is valid.
-	// If invalid, a non-nil error will be returned.
-	Validate() error
-}
-
-// NewServer creates a new admission webhook server.
-func NewServer(name string, mgr manager.Manager, options ServerOptions) (*Server, error) {
-	as := &Server{
-		Name:          name,
-		sMux:          http.NewServeMux(),
-		registry:      map[string]Webhook{},
-		ServerOptions: options,
-		manager:       mgr,
+// setDefaults does defaulting for the Server.
+func (s *Server) setDefaults() {
+	s.webhooks = map[string]http.Handler{}
+	s.webhookMux = http.NewServeMux()
+	if s.Port <= 0 {
+		s.Port = 443
+	}
+	if len(s.CertDir) == 0 {
+		s.CertDir = path.Join("/tmp", "k8s-webhook-server", "serving-certs")
 	}
-
-	return as, nil
 }
 
-// Register validates and registers webhook(s) in the server
-func (s *Server) Register(webhooks ...Webhook) error {
-	for i, webhook := range webhooks {
-		// validate the webhook before registering it.
-		err := webhook.Validate()
-		if err != nil {
-			return err
-		}
-		_, found := s.registry[webhook.GetPath()]
-		if found {
-			return fmt.Errorf("can't register duplicate path: %v", webhook.GetPath())
-		}
-		s.registry[webhook.GetPath()] = webhooks[i]
-		s.sMux.Handle(webhook.GetPath(), webhook.Handler())
+// Register marks the given webhook as being served at the given path.
+// It panics if two hooks are registered on the same path.
+func (s *Server) Register(path string, hook http.Handler) {
+	s.defaultingOnce.Do(s.setDefaults)
+	_, found := s.webhooks[path]
+	if found {
+		panic(fmt.Errorf("can't register duplicate path: %v", path))
 	}
-
-	// Lazily add Server to manager.
-	// Because the all webhook handlers to be in place, so we can inject the things they need.
-	return s.manager.Add(s)
+	s.webhooks[path] = hook
+	s.webhookMux.Handle(path, instrumentedHook(path, hook))
 }
 
-// Handle registers a http.Handler for the given pattern.
-func (s *Server) Handle(pattern string, handler http.Handler) {
-	s.sMux.Handle(pattern, handler)
-}
+// instrumentedHook adds some instrumentation on top of the given webhook.
+func instrumentedHook(path string, hookRaw http.Handler) http.Handler {
+	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+		startTS := time.Now()
+		defer metrics.RequestLatency.WithLabelValues(path).Observe(time.Now().Sub(startTS).Seconds())
```

Comment:
We need to wrap it with func{} otherwise the args `time.Now().Sub(startTS).Seconds()` will be evaluated immediately reporting almost no latency :)

Reply:
Good catch!
`The deferred call's arguments are evaluated immediately, but the function call is not executed until the surrounding function returns.`
Ref: https://tour.golang.org/flowcontrol/12

Reply:
ack, yep